packages/next/lib/recursive-delete.ts

```diff
@@ -0,0 +1,60 @@
+import fs from 'fs'
+import { join } from 'path'
+import { promisify } from 'util'
+
+const readdir = promisify(fs.readdir)
+const stat = promisify(fs.stat)
+const rmdir = promisify(fs.rmdir)
+const unlink = promisify(fs.unlink)
+const sleep = promisify(setTimeout)
+
+const unlinkFile = async (p: string, t = 1): Promise<void> => {
+  try {
+    await unlink(p)
+  } catch (e) {
+    if ((e.code === 'EBUSY' || e.code === 'ENOTEMPTY' || e.code === 'EPERM' || e.code === 'EMFILE') && t < 3) {
+      await sleep(t * 100)
+      return unlinkFile(p, t++)
+    }
+
+    if (e.code === 'ENOENT') {
+      return
+    }
+
+    throw e
+  }
+}
+
+/**
+ * Recursively delete directory contents
+ * @param  {string} dir Directory to delete the contents of
+ * @param  {RegExp} filter Filter for the file name, only the name part is considered, not the full path
+ * @returns Promise void
+ */
+export async function recursiveDelete(dir: string, filter?: RegExp, ensure?: boolean): Promise<void> {
+  let result
+  try {
+    result = await readdir(dir)
+  } catch (e) {
+    if (e.code === 'ENOENT' && !ensure) return
+
+    throw e
+  }
+
+  await Promise.all(result.map(async (part: string) => {
+    const absolutePath = join(dir, part)
+    const pathStat = await stat(absolutePath).catch((e) => {
+      if (e.code !== 'ENOENT') throw e
```

Comment:
Why is `ensure` checked above but not here? Can we add a comment in the code explaining that we only care to `ensure` directories and not files if that's the intended behavior?
```suggestion
      // We intentionally ignore missing files (unlike directories) because ...
      if (e.code !== 'ENOENT') throw e
```

Reply:
`ensure` just makes it so we can validate the the directory that's being targeted exists

Reply:
n.b. this is now documented in the `@param` comment for this arg